test/jit/test_isinstance.py

```diff
@@ -0,0 +1,246 @@
+import os
+import sys
+
+import torch
+from typing import List, Any, Dict, Tuple, Optional
+
+# Make the helper files in test/ importable
+pytorch_test_dir = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+sys.path.append(pytorch_test_dir)
+from torch.testing._internal.jit_utils import JitTestCase
+
+if __name__ == "__main__":
+    raise RuntimeError(
+        "This test file is not meant to be run directly, use:\n\n"
+        "\tpython test/test_jit.py TESTNAME\n\n"
+        "instead."
+    )
+
+# Tests for torch.jit.isinstance
+class TestIsinstance(JitTestCase):
+    def test_int(self):
+        def int_test(x: Any):
+            assert torch.jit.isinstance(x, int)
+            assert not torch.jit.isinstance(x, float)
+
+        x = 1
+        self.checkScript(int_test, (x,))
+
+    def test_float(self):
+        def float_test(x: Any):
+            assert torch.jit.isinstance(x, float)
+            assert not torch.jit.isinstance(x, int)
+
+        x = 1.0
+        self.checkScript(float_test, (x,))
+
+    def test_bool(self):
+        def bool_test(x: Any):
+            assert torch.jit.isinstance(x, bool)
+            assert not torch.jit.isinstance(x, float)
+
+        x = False
+        self.checkScript(bool_test, (x,))
+
+    def test_list(self):
+        def list_str_test(x: Any):
+            assert torch.jit.isinstance(x, List[str])
+            assert not torch.jit.isinstance(x, List[int])
+
+        x = ["1", "2", "3"]
+        self.checkScript(list_str_test, (x,))
+
+    def test_dict(self):
+        def dict_str_int_test(x: Any):
+            assert torch.jit.isinstance(x, Dict[str, int])
+            assert not torch.jit.isinstance(x, Dict[int, str])
+
+        x = {"a": 1, "b": 2}
+        self.checkScript(dict_str_int_test, (x,))
+
+    def test_tuple(self):
+        def tuple_test(x: Any):
+            assert torch.jit.isinstance(x, Tuple[str, int, str])
+            assert not torch.jit.isinstance(x, Tuple[int, str, str])
+            assert not torch.jit.isinstance(x, Tuple[str])
+
+        x = ("a", 1, "b")
+        self.checkScript(tuple_test, (x,))
+
+    def test_optional(self):
+        def optional_test(x: Any):
+            assert torch.jit.isinstance(x, Optional[torch.Tensor])
+            assert not torch.jit.isinstance(x, Optional[str])
+            # TODO: successful torch.jit.isinstance makes sets type?
+
+        x = torch.ones(3, 3)
+        self.checkScript(optional_test, (x,))
+
+    def test_optional_none(self):
+        def optional_test_none(x: Any):
+            assert torch.jit.isinstance(x, Optional[torch.Tensor])
+            # assert not torch.jit.isinstance(x, Optional[str])
+            # TODO: above line fails in TS interpreter need to investigate
```

Comment:
Are you looking into this?

Reply:
I'm running into the difficulty where eager mode and scripting behave differently. Eager mode will allow both isinstances to return true (since the None can be either in Optional[torch.Tensor] or Optional[str] Python), but from what I can tell during scripting type refinement happens during the first isinstance call refining the type to Optional[torch.Tensor] which then makes the 2nd isinstance call return false since the type isn't Optional[str]. 

I'm not sure how to go about resolving this. I don't know how to in python eager mode carry the type assignment outcome between isinstance calls to make the 2nd isinstance call return false

Reply:
Good finding.

I don't think we should change eager mode behavior as we would want users to mechanically migrate from `isinstance` to `torch.jit.isinstance`.

I think what we can do though is to change behavior of `prim::isinstance` in TorchScript. From what I can tell, `prim::isinstance` ultimate calls into [here](https://github.com/pytorch/pytorch/blob/38c97fb6f0725fa0c02687023b101f0f8339cca7/torch/csrc/jit/runtime/vararg_functions.cpp#L106) which checks for sub-typing relationship. I think this logic is only correct when all types either completely overlaps (subtype relationship) or do not overlap (not an instance of). `Optional[T]` doesn't fall into this situation. It would be even worse for `Union[T1, T2, T3]` when we support it in the future.

We should therefore fix this by changing implementation of `prim::isinstance`. It doesn't have to be in this diff though since this case is rare and unconventional enough. Feel free to leave a TODO here for now and we can discuss how to deal with it with more audience.